x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,421 @@
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.transport.NetworkExceptionHelper;
+import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.node.NodeClosedException;
 import org.elasticsearch.persistent.AllocatedPersistentTask;
 import org.elasticsearch.tasks.Task;
 import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.transport.ActionTransportException;
+import org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsAction;
+import org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsRequest;
+import org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsResponse;
 
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.LinkedList;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.LongConsumer;
 
+/**
+ * The node task that fetch the write operations from a leader shard and
+ * persists these ops in the follower shard.
+ */
 public class ShardFollowNodeTask extends AllocatedPersistentTask {
 
-    private final AtomicLong processedGlobalCheckpoint = new AtomicLong();
+    static final int DEFAULT_MAX_READ_SIZE = 1024;
+    static final int DEFAULT_MAX_WRITE_SIZE = 1024;
+    static final int RETRY_LIMIT = 10;
+    static final int DEFAULT_MAX_CONCURRENT_READS = 1;
+    static final int DEFAULT_MAX_CONCURRENT_WRITES = 1;
+    static final int DEFAULT_MAX_BUFFER_SIZE = 10240;
+    static final long DEFAULT_MAX_TRANSLOG_BYTES= Long.MAX_VALUE;
+    private static final TimeValue RETRY_TIMEOUT = TimeValue.timeValueMillis(500);
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardFollowNodeTask.class);
+
+    final Client leaderClient;
+    final Client followerClient;
+    private final ShardFollowTask params;
+    private final BiConsumer<TimeValue, Runnable> scheduler;
+
+    private volatile long lastRequestedSeqno;
+    private volatile long leaderGlobalCheckpoint;
+
+    private volatile int numConcurrentReads = 0;
+    private volatile int numConcurrentWrites = 0;
+    private volatile long processedGlobalCheckpoint = 0;
```

Comment:
those variables seem to be read together. I wonder if we should not make the volatile but rather synchronize their reading?

Reply:
Not sure. I don't want to make the `getStatus()` call synchronized which can be invoked when list tasks api is used.